File: 1359.count-all-valid-pickup-and-delivery-options.cpp
/*
 * @lc app=leetcode id=1359 lang=cpp
 *
 * [1359] Count All Valid Pickup and Delivery Options
 */

// @lc code=start
class Solution {
public:
    static constexpr int MODULO = 1e9+7;
    int ModuloMultiply(int a, int b) {
        return ((int64_t)(a % MODULO)) * ((int64_t)(b % MODULO)) % MODULO;
    }
    int countOrders(int n) {
        int options[n+1];
        options[1] = 1;
        for (int i = 2; i <= n; i++) {
            options[i] = ModuloMultiply(ModuloMultiply(options[i-1], i), 2*i-1);
        }
        return options[n];
    }
};
// @lc code=end

File: 1359.count-all-valid-pickup-and-delivery-options.rs
/*
 * @lc app=leetcode id=1359 lang=rust
 *
 * [1359] Count All Valid Pickup and Delivery Options
 */

// @lc code=start
impl Solution {
    pub fn modulo_multiply(a : i32, b : i32) -> i32 {
        let p : i64 = 1e9 as i64 + 7;
        return ((a as i64 % p) * (b as i64 % p) % p) as i32;
    }
    pub fn count_orders(n: i32) -> i32 {
        let mut prev = 1;
        let mut current = 1;
        for i in 2..=n {
            current = Solution::modulo_multiply(Solution::modulo_multiply(prev, i), (2 * i - 1));
            prev = current;
        }
        return current;
    }
}
// @lc code=end

